File: packages/editor/src/phrase.ts
import { assignFingerings, type FingeringConstraints } from './fingering'
import { availableFretCount, soundingTuning } from './instrument'
import type { FingeringMode, NoteEvent, TabDocument } from './types'

export type PhrasePlanErrorCode =
  | 'invalid-range'
  | 'invalid-constraints'
  | 'empty-range'
  | 'impossible-constraints'

export class PhrasePlanError extends Error {
  readonly code: PhrasePlanErrorCode

  constructor(code: PhrasePlanErrorCode, message: string, options?: ErrorOptions) {
    super(message, options)
    this.name = 'PhrasePlanError'
    this.code = code
  }
}

export interface PhraseScoreRange {
  startScoreTick: number
  endScoreTick: number
}

export interface PhraseFingeringRequest {
  range: PhraseScoreRange
  mode: FingeringMode
  constraints?: FingeringConstraints
}

export interface PhraseFingeringChange {
  noteId: string
  scoreTick: number
  midiPitch: number
  before: { string: number; fret: number }
  after: { string: number; fret: number }
}

export interface PhraseFingeringPlan {
  range: PhraseScoreRange
  mode: FingeringMode
  constraints: FingeringConstraints
  notes: NoteEvent[]
  changes: PhraseFingeringChange[]
  selectedNoteCount: number
  lockedNoteCount: number
}

function barTicks(tab: TabDocument): number {
  const [beats, beatUnit] = tab.time_signature
  const ticks = (tab.ticks_per_quarter * 4 * beats) / beatUnit
  if (!Number.isInteger(ticks) || ticks <= 0) {
    throw new PhrasePlanError(
      'invalid-range',
      'Time signature does not divide into whole score ticks.',
    )
  }
  return ticks
}

function normalizedConstraints(
  tab: TabDocument,
  constraints: FingeringConstraints | undefined,
): FingeringConstraints {
  const allowedStrings = constraints?.allowedStrings
    ? [...new Set(constraints.allowedStrings)].sort((left, right) => left - right)
    : undefined
  const minFret = constraints?.minFret ?? null
  const maxFret = constraints?.maxFret ?? null
  if (
    allowedStrings?.some(
      (string) => !Number.isInteger(string) || string < 1 || string > tab.tuning.length,
    ) ||
    (allowedStrings && allowedStrings.length === 0) ||
    (minFret != null && (!Number.isInteger(minFret) || minFret < 0)) ||
    (maxFret != null && (!Number.isInteger(maxFret) || maxFret > availableFretCount(tab))) ||
    (minFret != null && maxFret != null && minFret > maxFret)
  ) {
    throw new PhrasePlanError(
      'invalid-constraints',
      'Choose at least one valid string and a fret range available on this instrument.',
    )
  }
  return { allowedStrings, minFret, maxFret }
}

function expandToBars(tab: TabDocument, requested: PhraseScoreRange): PhraseScoreRange {
  if (
    !Number.isFinite(requested.startScoreTick) ||
    !Number.isFinite(requested.endScoreTick) ||
    requested.startScoreTick < 0 ||
    requested.endScoreTick <= requested.startScoreTick
  ) {
    throw new PhrasePlanError(
      'invalid-range',
      'Phrase range must have a non-negative start before its end.',
    )
  }
  const ticksPerBar = barTicks(tab)
  const offset = (tab as TabDocument & { bar_offset_ticks?: number }).bar_offset_ticks ?? 0
  const startScoreTick = Math.max(
    0,
    Math.floor((requested.startScoreTick - offset) / ticksPerBar) * ticksPerBar + offset,
  )
  const endScoreTick =
    Math.ceil((requested.endScoreTick - offset) / ticksPerBar) * ticksPerBar + offset
  return { startScoreTick, endScoreTick }
}

function groupIndices(notes: readonly NoteEvent[], start: number, step: -1 | 1): number[] {
  const neighbor = notes[start]
  if (!neighbor) return []
  const indices = [start]
  for (let index = start + step; notes[index]?.onset_frame === neighbor.onset_frame; index += step) {
    indices.push(index)
  }
  return indices
}

/**
 * Plan a bar-aligned partial refingering. Input tab stays untouched.
 * Neighboring onset groups are locked context, so hand movement and connected
 * techniques remain smooth across both phrase boundaries.
 */
export function planPhraseFingering(
  tab: TabDocument,
  request: PhraseFingeringRequest,
): PhraseFingeringPlan {
  const range = expandToBars(tab, request.range)
  const constraints = normalizedConstraints(tab, request.constraints)
  const chronological = tab.notes
    .map((note, originalIndex) => ({ note, originalIndex }))
    .sort(
      (left, right) =>
        left.note.onset_frame - right.note.onset_frame ||
        left.originalIndex - right.originalIndex,
    )
  const selected = chronological
    .map(({ note }, index) => ({ note, index }))
    .filter(
      ({ note }) =>
        note.score_tick >= range.startScoreTick && note.score_tick < range.endScoreTick,
    )
  if (selected.length === 0) {
    throw new PhrasePlanError(
      'empty-range',
      'Selected bars contain no notes. Choose bars containing tablature.',
    )
  }

  const selectedIndices = new Set(selected.map(({ index }) => index))
  const firstIndex = selected[0].index
  const lastIndex = selected.at(-1)!.index
  const contextIndices = new Set([
    ...groupIndices(chronological.map(({ note }) => note), firstIndex - 1, -1),
    ...groupIndices(chronological.map(({ note }) => note), lastIndex + 1, 1),
  ])
  const workingIndices = [...new Set([...selectedIndices, ...contextIndices])].sort(
    (left, right) => left - right,
  )
  const workingNotes = workingIndices.map((index) => {
    const note = chronological[index].note
    return contextIndices.has(index) ? { ...note, user_locked: true } : note
  })

  let plannedWorking: NoteEvent[]
  try {
    plannedWorking = assignFingerings(
      workingNotes,
      soundingTuning(tab),
      availableFretCount(tab),
      request.mode,
      tab.preferred_fret,
      constraints,
    )
  } catch (error) {
    throw new PhrasePlanError(
      'impossible-constraints',
      error instanceof Error
        ? `No playable fingering fits these phrase controls. ${error.message}`
        : 'No playable fingering fits these phrase controls.',
      { cause: error },
    )
  }

  const notes = tab.notes.map((note) => ({ ...note }))
  workingIndices.forEach((chronologicalIndex, workingIndex) => {
    if (!selectedIndices.has(chronologicalIndex)) return
    const originalIndex = chronological[chronologicalIndex].originalIndex
    const before = tab.notes[originalIndex]
    const planned = plannedWorking[workingIndex]
    const changed = before.string !== planned.string || before.fret !== planned.fret
    notes[originalIndex] = changed
      ? { ...planned, reviewed: false, user_locked: false }
      : { ...planned, reviewed: before.reviewed, user_locked: before.user_locked }
  })
  const changes = tab.notes.flatMap((before, index): PhraseFingeringChange[] => {
    const after = notes[index]
    if (before.string === after.string && before.fret === after.fret) return []
    return [{
      noteId: before.id,
      scoreTick: before.score_tick,
      midiPitch: before.midi_pitch,
      before: { string: before.string, fret: before.fret },
      after: { string: after.string, fret: after.fret },
    }]
  })
  return {
    range,
    mode: request.mode,
    constraints,
    notes,
    changes,
    selectedNoteCount: selected.length,
    lockedNoteCount: selected.filter(({ note }) => note.user_locked).length,
  }
}

File: server/solotrace/phrase.py
from __future__ import annotations

import math
from dataclasses import dataclass

from .fingering import FingeringConstraints, FingeringMode, assign_fingerings
from .models import NoteEvent, TabDocument


class PhrasePlanError(ValueError):
    def __init__(self, code: str, message: str) -> None:
        super().__init__(message)
        self.code = code


@dataclass(frozen=True)
class PhraseScoreRange:
    start_score_tick: int
    end_score_tick: int


@dataclass(frozen=True)
class PhraseFingeringChange:
    note_id: str
    score_tick: int
    midi_pitch: int
    before_string: int
    before_fret: int
    after_string: int
    after_fret: int


@dataclass(frozen=True)
class PhraseFingeringPlan:
    range: PhraseScoreRange
    mode: FingeringMode
    constraints: FingeringConstraints
    notes: list[NoteEvent]
    changes: list[PhraseFingeringChange]
    selected_note_count: int
    locked_note_count: int


def _bar_ticks(tab: TabDocument) -> int:
    beats, beat_unit = tab.time_signature
    ticks = tab.ticks_per_quarter * 4 * beats / beat_unit
    if not ticks.is_integer() or ticks <= 0:
        raise PhrasePlanError(
            "invalid-range",
            "Time signature does not divide into whole score ticks.",
        )
    return int(ticks)


def _expanded_range(
    tab: TabDocument,
    start_score_tick: int,
    end_score_tick: int,
) -> PhraseScoreRange:
    if start_score_tick < 0 or end_score_tick <= start_score_tick:
        raise PhrasePlanError(
            "invalid-range",
            "Phrase range must have a non-negative start before its end.",
        )
    ticks_per_bar = _bar_ticks(tab)
    offset = getattr(tab, "bar_offset_ticks", 0)
    start = max(
        0,
        math.floor((start_score_tick - offset) / ticks_per_bar) * ticks_per_bar
        + offset,
    )
    end = (
        math.ceil((end_score_tick - offset) / ticks_per_bar) * ticks_per_bar
        + offset
    )
    return PhraseScoreRange(start_score_tick=start, end_score_tick=end)


def _validated_constraints(
    tab: TabDocument,
    constraints: FingeringConstraints | None,
) -> FingeringConstraints:
    candidate = constraints or FingeringConstraints()
    strings = candidate.allowed_strings
    if (
        (
            strings is not None
            and (not strings or any(s < 1 or s > len(tab.tuning) for s in strings))
        )
        or (candidate.min_fret is not None and candidate.min_fret < 0)
        or (candidate.max_fret is not None and candidate.max_fret > tab.available_fret_count)
        or (
            candidate.min_fret is not None
            and candidate.max_fret is not None
            and candidate.min_fret > candidate.max_fret
        )
    ):
        raise PhrasePlanError(
            "invalid-constraints",
            "Choose at least one valid string and a fret range available on this instrument.",
        )
    return candidate


def _group_indices(notes: list[NoteEvent], start: int, step: int) -> list[int]:
    if not 0 <= start < len(notes):
        return []
    onset = notes[start].onset_frame
    indices: list[int] = []
    index = start
    while 0 <= index < len(notes) and notes[index].onset_frame == onset:
        indices.append(index)
        index += step
    return indices


def plan_phrase_fingering(
    tab: TabDocument,
    *,
    start_score_tick: int,
    end_score_tick: int,
    mode: FingeringMode,
    constraints: FingeringConstraints | None = None,
) -> PhraseFingeringPlan:
    """Plan bar-aligned partial fingering without mutating the source tab."""
    range_ = _expanded_range(tab, start_score_tick, end_score_tick)
    normalized_constraints = _validated_constraints(tab, constraints)
    chronological = sorted(
        enumerate(tab.notes),
        key=lambda indexed: (indexed[1].onset_frame, indexed[0]),
    )
    arranged = [note for _, note in chronological]
    selected_indices = {
        index
        for index, note in enumerate(arranged)
        if range_.start_score_tick <= note.score_tick < range_.end_score_tick
    }
    if not selected_indices:
        raise PhrasePlanError(
            "empty-range",
            "Selected bars contain no notes. Choose bars containing tablature.",
        )

    first_index = min(selected_indices)
    last_index = max(selected_indices)
    context_indices = {
        *_group_indices(arranged, first_index - 1, -1),
        *_group_indices(arranged, last_index + 1, 1),
    }
    working_indices = sorted(selected_indices | context_indices)
    working_notes = [
        arranged[index].model_copy(update={"user_locked": True})
        if index in context_indices
        else arranged[index]
        for index in working_indices
    ]
    try:
        planned_working = assign_fingerings(
            working_notes,
            tab.sounding_tuning,
            tab.available_fret_count,
            mode,
            tab.preferred_fret,
            normalized_constraints,
        )
    except ValueError as error:
        raise PhrasePlanError(
            "impossible-constraints",
            f"No playable fingering fits these phrase controls. {error}",
        ) from error

    notes = [note.model_copy(deep=True) for note in tab.notes]
    for working_index, chronological_index in enumerate(working_indices):
        if chronological_index not in selected_indices:
            continue
        original_index = chronological[chronological_index][0]
        before = tab.notes[original_index]
        planned = planned_working[working_index]
        changed = (before.string, before.fret) != (planned.string, planned.fret)
        notes[original_index] = (
            planned.model_copy(update={"reviewed": False, "user_locked": False})
            if changed
            else before.model_copy(deep=True)
        )

    changes = [
        PhraseFingeringChange(
            note_id=before.id,
            score_tick=before.score_tick,
            midi_pitch=before.midi_pitch,
            before_string=before.string,
            before_fret=before.fret,
            after_string=after.string,
            after_fret=after.fret,
        )
        for before, after in zip(tab.notes, notes, strict=True)
        if (before.string, before.fret) != (after.string, after.fret)
    ]
    selected_notes = [arranged[index] for index in selected_indices]
    return PhraseFingeringPlan(
        range=range_,
        mode=mode,
        constraints=normalized_constraints,
        notes=notes,
        changes=changes,
        selected_note_count=len(selected_indices),
        locked_note_count=sum(note.user_locked for note in selected_notes),
    )
